torch/distributed/distributed_c10d.py

```diff
@@ -2797,3 +2791,124 @@ def new_group(ranks=None, timeout=default_pg_timeout, backend=None, pg_options=N
+            all the output subgroups. If None, the default process group will be used.
+            Default is ``None``.
+        timeout (timedelta, optional): Timeout for operations executed against
+            the process group. Default value equals 30 minutes.
+            This is applicable for the ``gloo`` backend. For ``nccl``, this is
+            applicable only if the environment variable ``NCCL_BLOCKING_WAIT``
+            or ``NCCL_ASYNC_ERROR_HANDLING`` is set to 1. When
+            ``NCCL_BLOCKING_WAIT`` is set, this is the duration for which the
+            process will block and wait for collectives to complete before
+            throwing an exception. When ``NCCL_ASYNC_ERROR_HANDLING`` is set,
+            this is the duration after which collectives will be aborted
+            asynchronously and the process will crash. ``NCCL_BLOCKING_WAIT``
+            will provide errors to the user which can be caught and handled,
+            but due to its blocking nature, it has a performance overhead. On
+            the other hand, ``NCCL_ASYNC_ERROR_HANDLING`` has very little
+            performance overhead, but crashes the process on errors. This is
+            done since CUDA execution is async and it is no longer safe to
+            continue executing user code since failed async NCCL operations
+            might result in subsequent CUDA operations running on corrupted
+            data. Only one of these two environment variables should be set.
+        backend (str or Backend, optional): The backend to use. Depending on
+            build-time configurations, valid values are ``gloo`` and ``nccl``.
+            By default uses the same backend as the global group. This field
+            should be given as a lowercase string (e.g., ``"gloo"``), which can
+            also be accessed via :class:`Backend` attributes (e.g.,
+            ``Backend.GLOO``). If ``None`` is passed in, the backend
+            corresponding to the default process group will be used. Default is
+            ``None``.
+        pg_options (ProcessGroupOptions, optional): process group options
+            specifying what additional options need to be passed in during
+            the construction of specific process groups. i.e. for the ``nccl``
+            backend, is_high_priority_stream can be specified so that process
+            group can pick up high priority cuda streams.
+
+    Returns:
+        The subgroup containing the current rank, and all the subgroups used for cleanup.
+    """
+    subgroups = []
+    cur_subgroup = None
+    if ranks_per_subgroup_list:
+        for ranks in ranks_per_subgroup_list:
+            subgroup = new_group(
+                ranks=ranks, timeout=timeout, backend=backend, pg_options=pg_options
+            )
+            subgroups.append(subgroup)
+
+            rank = get_rank()
+            if rank in ranks:
+                cur_subgroup = subgroup
```

Comment:
What if there is an overlap? It looks like we return the last subgroup that the rank is a part of, should we return all of them somehow? Or explicitly forbid overlaps?

Reply:
Good question. Actually we can support overlapping subgroups. Here since we want to return a current subgroup, overlap is forbidden. Otherwise, there can be multiple current subgroups.

I don't see such a need at this time, so let's keep things simple.